imap_processing/ultra/l1c/ultra_l1c_pset_bins.py

```diff
@@ -44,11 +48,221 @@ def build_spatial_bins(spacing: float = 0.5) -> tuple[np.ndarray, np.ndarray]:
+    az_bin_id, el_bin_id, energy_bin_id = np.argwhere(hist > 0).T
+
+    # Unique identifier for spatial bins by combining azimuth and elevation indices
+    bin_id = (np.unique(az_bin_id), np.unique(el_bin_id), np.unique(energy_bin_id))
+
+    # Get unique azimuth and elevation midpoints
+    midpoints = (
+        np.unique(az_bin_midpoints[az_bin_id]),
+        np.unique(el_bin_midpoints[el_bin_id]),
+    )
+
+    # Determine positions in unique coordinates for assigning counts
+    az_positions = np.searchsorted(bin_id[0], az_bin_id)
+    el_positions = np.searchsorted(bin_id[1], el_bin_id)
+    energy_positions = np.searchsorted(bin_id[2], energy_bin_id)
+
+    counts = np.zeros((len(bin_id[0]), len(bin_id[1]), len(bin_id[2])))
+    counts[az_positions, el_positions, energy_positions] = non_zero_counts
+
+    return counts, bin_id, midpoints
+
+
+def bin_data(v: tuple[np.ndarray, np.ndarray, np.ndarray], energy: np.ndarray) -> tuple:
+    """
+    Bin the data spatially and by energy.
+
+    Parameters
+    ----------
+    v : tuple[np.ndarray, np.ndarray, np.ndarray]
+        A 2D array containing the x, y, z components of velocity vector.
+    energy : np.ndarray
+        A 1D array with shape (N,) containing the particle energy.
+
+    Returns
+    -------
+    counts : np.ndarray
+        A 2D array containing the binned counts for each spatial and energy bin.
+    bin_id : tuple[np.ndarray, np.ndarray]
+        Unique identifiers for the spatial and energy bins.
+    midpoints : tuple[np.ndarray, np.ndarray]
+        Azimuth and elevation midpoints for the spatial bins.
+    energy_edge_start : np.ndarray
+        The starting edges of the energy bins corresponding to non-zero counts.
+
+    Notes
+    -----
+    The returned data can be used to construct a xarray Dataset as follows:
+
+    ds = xr.Dataset(
+        {
+            "az_midpoint": ("az_bin_id", midpoints[0]),
+            "el_midpoint": ("el_bin_id", midpoints[1]),
+            "energy_edge_start": ("energy_bin_id", energy_edge_start),
+            "counts": (("az_bin_id", "el_bin_id","energy_bin_id"), counts),
+        },
+        coords={
+            "az_bin_id": bin_id[0],
+            "el_bin_id": bin_id[1],
+            "energy_bin_id": bin_id[2],
+        }
+    )
+    """
+    az_bin_edges, el_bin_edges, az_bin_midpoints, el_bin_midpoints = (
+        build_spatial_bins()
+    )
+    energy_bin_edges = build_energy_bins()
+
+    # Compute the 3D histogram of the particle data
+    hist, non_zero_counts = get_histogram(
+        v, energy, az_bin_edges, el_bin_edges, energy_bin_edges
+    )
+
+    # Get the starting edges for energy bins
+    energy_bin_start = energy_bin_edges[:-1]
+
+    # Create unique identifiers and map non-zero counts to 2D histogram array
+    counts, bin_id, midpoints = create_unique_identifiers(
+        hist, az_bin_midpoints, el_bin_midpoints, non_zero_counts
+    )
+
+    _, _, energy_bin = np.argwhere(hist > 0).T
```

Comment:
I see this use of `np.argwhere(hist > 0).T` in 2 places and also `non_zero_counts = hist[hist>0]` in another place. It would be good to consolidate those to a single place so that there aren't hidden inter-dependencies between these three new functions.